crates/re_types/src/lib.rs

```diff
@@ -109,6 +109,73 @@
 
 // ---
 
+/// Describes the interface for interpreting an object as a bundle of [`Component`]s.
+pub trait AsComponents {
+    /// Exposes the object's contents as a set of [`ComponentBatch`]s.
+    ///
+    /// This is the main mechanism for easily extending builtin archetypes or even writing
+    /// fully custom ones.
+    /// Have a look at our [Custom Data] example to learn more about extending archetypes.
+    ///
+    /// [Custom Data]: https://github.com/rerun-io/rerun/blob/latest/examples/rust/custom_data/src/main.rs
+    //
+    // NOTE: Don't bother returning a CoW here: we need to dynamically discard optional components
+    // depending on their presence (or lack thereof) at runtime anyway.
+    fn as_component_batches(&self) -> Vec<MaybeOwnedComponentBatch<'_>>;
+
+    /// The number of instances in each batch.
+    ///
+    /// If not implemented, the number of instances will be determined by the longest
+    /// batch in the bundle.
+    ///
+    /// Each batch returned by `as_component_batches` should have this number of
+    /// elements, or 1 in the case it is a splat, or 0 in the case that
+    /// component is being cleared.
+    #[inline]
+    fn num_instances(&self) -> usize {
+        self.as_component_batches()
+            .first()
+            .map_or(0, |comp_batch| comp_batch.as_ref().num_instances())
```

Comment:
docstring and implementation differs

Taking the max seems a beeter choice here - there is nothing in the docs for `as_component_batches` that mandates that the first component need to be special somehow.